Scheduling Profiler: Redesign with DevTools styling

taneliang opened the pull request:
<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.

  Before submitting a pull request, please make sure the following is done:

  1. Fork [the repository](https://github.com/facebook/react) and create your branch from `master`.
  2. Run `yarn` in the repository root.
  3. If you've fixed a bug or added code that should be tested, add tests!
  4. Ensure the test suite passes (`yarn test`). Tip: `yarn test --watch TestName` is helpful in development.
  5. Run `yarn test-prod` to test in the production environment. It supports the same options as `yarn test`.
  6. If you need a debugger, run `yarn debug-test --watch TestName`, open `chrome://inspect`, and press "Inspect".
  7. Format your code with [prettier](https://github.com/prettier/prettier) (`yarn prettier`).
  8. Make sure your code lints (`yarn lint`). Tip: `yarn linc` to only check changed files.
  9. Run the [Flow](https://flowtype.org/) type checks (`yarn flow`).
  10. If you haven't already, complete the CLA.

  Learn more about contributing: https://reactjs.org/docs/how-to-contribute.html
-->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

Adopt DevTools styling in the scheduling profiler and show a persistent toolbar at the top of the page.

![image](https://user-images.githubusercontent.com/12784593/91420425-48ede280-e887-11ea-8b68-06a1649e6f76.png)

Resolves #19676.

cc @bvaughn, @kartikcho

### High-Level Goals

Design: Match React DevTools's design, so that we carry the same brand and level of professionalism.

Code: Revamp `App` code structure and CSS to resemble the `DevTools` component -- this is intended to make it slightly easier to merge the scheduling profiler into DevTools when the time comes.

Future expansion: Create space on the canvas page to display the React version (with a view towards #19668) and other profile metadata e.g. profile name and browser type.

Misc features:

- Display an import button even after a profile has been loaded.
- Add an import error dialog.

### Implementation Breakdown

This is a large PR, but a lot of the code is copied verbatim from `react-devtools-shared`.

- Replaces the existing `ImportPage`, mainly with `ImportButton`
- Fork these from `react-devtools-shared` unchanged: Button, ButtonIcon, Icon, ModalDialog, ReactLogo, TabBar, Tooltip, root.css, utils/storage.js
- Fork these from `react-devtools-shared` with modifications:
	- Settings: I've modified this to remove its dependency on BridgeContext and StoreContext, both of which are not used in the scheduling profiler. I've left the list of CSS variables unchanged even though most of them aren't used as far as I can tell.
	- hooks.js: Removed unused hooks and add a `useBrowserTheme` hook for auto dark mode 🎉 
- Inject a new `DEVTOOLS_VERSION` variable similar to DevTools', but computed from the scheduling profiler's package.json instead.
- Overhaul `App.js` with the rough component hierarchy and same CSS styles as `DevTools.js`
- Create a new `SchedulingProfiler` component following `Profiler.js`.

If we decide to accept this PR, we'll want to adapt #19690's help text to the new design and code structure.

### Questions

- Should we import code from `react-devtools-shared` instead of forking their implementations?
- Are there any standard sources for the DevTools icons? I did a cursory search and found an icon in the Chromium source code, but I'm not sure where the others are from.
- If we want to continue forking the implementations, should we delete unused code or leave them around? I'd imagine that leaving them around will ease the eventual merging as the diff will be minimized.

### TODO

- [x] Resolve questions above
- [x] Place dark mode behind a new feature flag; the profiler canvas does not support dark mode. We'll leave comfortable sizing there; even though the canvas doesn't support comfortable sizing, I think people may appreciate having the rest of the UI text scale.

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes the user interface. -->

Before import (equivalent to the old `ImportPage`):

| Before | After |
|-|-|
|![image](https://user-images.githubusercontent.com/12784593/91420464-530fe100-e887-11ea-8279-df7ed9f34f44.png)|![image](https://user-images.githubusercontent.com/12784593/91426519-35df1080-e88f-11ea-86fb-6fd3761541d1.png)|

After import:

| Before | After |
|-|-|
|![image](https://user-images.githubusercontent.com/12784593/91423563-4b523b80-e88b-11ea-938a-95c9ce811b66.png)|![image](https://user-images.githubusercontent.com/12784593/91423495-32e22100-e88b-11ea-85ef-98368f33fa8b.png)|

Because we're using DevTools code, there's built-in support for configurable (auto!) dark mode and display density. Unfortunately, this is not very useful now because the profiler canvas doesn't have support for CSS variables yet. I'll hide it behind a feature flag, but this is what it looks like now:

Settings page:

![image](https://user-images.githubusercontent.com/12784593/91423307-f6aec080-e88a-11ea-9403-a7ca3c8e9fbd.png)

Comfortable dark:

![image](https://user-images.githubusercontent.com/12784593/91423335-ff06fb80-e88a-11ea-8d5a-2267d3264e0d.png)

Compact dark:

![image](https://user-images.githubusercontent.com/12784593/91423357-062e0980-e88b-11ea-808c-db712cc1d908.png)

Compact dark:

![image](https://user-images.githubusercontent.com/12784593/91423416-13e38f00-e88b-11ea-9231-3df92da27221.png)

Comfortable dark:

![image](https://user-images.githubusercontent.com/12784593/91423811-9cfac600-e88b-11ea-8329-8189ab1e76ca.png)

Hidden image when viewport is shorter than 600px:

![image](https://user-images.githubusercontent.com/12784593/91423858-aa17b500-e88b-11ea-8cc2-2519dba63f42.png)

New import failure dialog:

![image](https://user-images.githubusercontent.com/12784593/91426019-83a74900-e88e-11ea-873a-957c79940084.png)

### Known Issues

It's possible to hover over the canvas content even if the settings dialog is up. I can fix that in a separate PR to prevent this PR from bloating further.

![image](https://user-images.githubusercontent.com/12784593/91425807-39be6300-e88e-11ea-9e70-ffd7171dfcd6.png)

bvaughn commented:
This is in my queue to review. I was on PTO last week though and I have a lot of catch up to do. I'll try to get to it sometime in the next couple of days. :smile:

There is a Yarn lockfile conflict that might be nice to fix when you get the chance though.

taneliang commented:
No problem, take your time! 😄

I've rebased on master and fixed the conflict, and made 2 tiny changes:

- Bumped the Reach packages to their latest versions
- Removed the version string from the toolbar as I doubt users will want/need to see it. The version number can still be found in the settings panel.
![image](https://user-images.githubusercontent.com/12784593/91990984-f5870300-ed64-11ea-9c02-5271565d13e1.png)

taneliang commented:
Also hid the dark mode setting behind a new scheduling profiler feature flag `enableDarkMode`.

App no longer goes into dark mode even if the app was previously set to Auto or Dark

![image](https://user-images.githubusercontent.com/12784593/91994480-397c0700-ed69-11ea-8b5e-2ea264c8a299.png)

Theme setting is now also hidden

![image](https://user-images.githubusercontent.com/12784593/91994541-48fb5000-ed69-11ea-8a9a-1ba5568a6e3f.png)

Dark mode still available when `enableDarkMode = true`

![image](https://user-images.githubusercontent.com/12784593/91994991-d76fd180-ed69-11ea-804a-edd33bd0f245.png)

taneliang commented:
I'm rebasing this on master

bvaughn commented:
> * Should we import code from react-devtools-shared instead of forking their implementations?

Importing seems nice!

> * Are there any standard sources for the DevTools icons? I did a cursory search and found an icon in the Chromium source code, but I'm not sure where the others are from.

Most of them come from Material UI: https://material-ui.com/components/material-icons/

> * If we want to continue forking the implementations, should we delete unused code or leave them around? I'd imagine that leaving them around will ease the eventual merging as the diff will be minimized.

No strong preference, but I also think using `react-devtools-shared` seems nice so maybe this question doesn't apply.

bvaughn commented on packages/react-devtools-scheduling-profiler/src/hooks.js:
```diff
@@ -0,0 +1,199 @@
+    };
+
+    const mediaQueryList = window.matchMedia('(prefers-color-scheme: dark)');
+    mediaQueryList.addEventListener('change', handlePreferredColorSchemeChange);
```
I don't think this picks up on dark mode if it's already set (only if the user changes their system theme *after* the profiler has been opened). At least that seems to be the behavior on OSX (the "change" event won't get fired).

We could *also* check `window.matchMedia(DARK_MODE_QUERY).matches` in the effect and update, but then we might have to render twice (once for the default "light" and then again for "dark").

I think this might be a nice opportunity to use the `useMutableSource` hook! :smile: It will handle all of this for us. Here's how you'd use it:

```js
import {
  unstable_createMutableSource as createMutableSource,
  unstable_useMutableSource as useMutableSource,
  useEffect,
  useState,
} from 'react';

export type BrowserTheme = 'dark' | 'light';

const DARK_MODE_QUERY = '(prefers-color-scheme: dark)';

const getSnapshot = window =>
  window.matchMedia(DARK_MODE_QUERY).matches ? 'dark' : 'light';

const darkModeMutableSource = createMutableSource(
  window,
  () => window.matchMedia(DARK_MODE_QUERY).matches,
);

const subscribe = (window, callback) => {
  const mediaQueryList = window.matchMedia(DARK_MODE_QUERY);
  mediaQueryList.addEventListener('change', callback);
  return () => {
    mediaQueryList.removeEventListener('change', callback);
  };
};

export function useBrowserTheme(): BrowserTheme {
  return useMutableSource(darkModeMutableSource, getSnapshot, subscribe);
}
```

taneliang replied:
Wow, cool! I didn't know this existed 😆

bvaughn replied:
It's not released yet in stable, but DevTools uses prerelease APIs :smile:

https://github.com/reactjs/rfcs/blob/master/text/0147-use-mutable-source.md

bvaughn commented on packages/react-devtools-scheduling-profiler/src/TabBar.js:
```diff
@@ -0,0 +1,133 @@
+/**
```
Let's use `react-devtools-shared/src/devtools/views/TabBar.js` and `react-devtools-shared/src/devtools/views/TabBar.css` rather than fork.

bvaughn reviewed (commented):
I'm going to sit down the review now since most of what I'm writing is "use the shared thing rather than fork" 😁  I'd be happy to make that change myself, or you could. Your call!

bvaughn commented:
Eh, screw it :smile: I'll hop in and help since it's after hours for you. Give me a few minutes and I'll push a commit that does my won nits.

bvaughn commented on packages/react-devtools-scheduling-profiler/src/App.js:
```diff
@@ -7,22 +7,32 @@
-import ImportPage from './ImportPage';
-import CanvasPage from './CanvasPage';
+import {ModalDialogContextController} from './ModalDialog';
+import {SettingsContextController} from './Settings/SettingsContext';
```
I'm not convinced the settings dialog is worth the hassle of the forked contexts and files. For now, I'm going to just rip it out :D

taneliang commented:
>  Give me a few minutes and I'll push a commit that implements my own nits.

Thanks! I gotta get back to my art history class readings 😆 Let me know if there's anything I can do! I'll be happy to follow up on anything tomorrow afternoon.

bvaughn reviewed (approved):
👍

bvaughn commented on packages/react-devtools-scheduling-profiler/src/hooks.js:
```diff
@@ -53,6 +52,8 @@ export function useBrowserTheme(): void {
           throw Error(`Unsupported theme value "${theme}"`);
       }
+    } else {
+      updateThemeVariables('light', documentElements);
```
@taneliang This should be `theme` not `'light'` right?

bvaughn replied:
Ah this was outside of the feature flag conditional. My apologies.

taneliang replied:
Oh I think we want `'light'` to prevent the app from going into dark mode?
